### PR TITLE
Fix string comparison with ordinal casing with Surrogates

### PR DESCRIPTION
**Pull request by tarekgh**
This change is fixing the string comparisons with ordinal casing when using surrogate pairs. 
I have measured the perf for ordinal casing comparisons in general to ensure there is no regression introduced because of this change.  

*.NET 6.0 Before changes*

```
|                             Method |      Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------------------- |----------:|----------:|----------:|------:|------:|------:|----------:|
|          SimpleAsciiNoCasingNeeded |  1.626 ns | 0.0603 ns | 0.0619 ns |     - |     - |     - |         - |
|            SimpleAsciiCasingNeeded | 31.667 ns | 0.3536 ns | 0.3135 ns |     - |     - |     - |         - |
|              NonSimpleCasingNeeded | 40.866 ns | 0.6449 ns | 0.9453 ns |     - |     - |     - |         - |
| NonSimpleCasingNeededWithSurrogate | 48.083 ns | 0.9228 ns | 0.8180 ns |     - |     - |     - |         - |
```

*.NET 6.0 After changes*

```
|                             Method |      Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------------------- |----------:|----------:|----------:|------:|------:|------:|----------:|
|          SimpleAsciiNoCasingNeeded |  1.309 ns | 0.0251 ns | 0.0222 ns |     - |     - |     - |         - |
|            SimpleAsciiCasingNeeded | 30.796 ns | 0.3370 ns | 0.2988 ns |     - |     - |     - |         - |
|              NonSimpleCasingNeeded | 40.400 ns | 0.5521 ns | 0.5164 ns |     - |     - |     - |         - |
| NonSimpleCasingNeededWithSurrogate | 48.030 ns | 0.7178 ns | 0.6715 ns |     - |     - |     - |         - |
```

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @safern 

No rush for this one :-)

File: src/libraries/System.Private.CoreLib/src/System/Globalization/OrdinalCasing.Icu.cs [hunk 1]
```diff
@@ -196,72 +197,90 @@ internal static int CompareStringIgnoreCase(ref char strA, int lengthA, ref char
...
+            int index = 0;
+
+            while (index < length)
             {
-                 // optimize for Ascii cases
-                if (charA <= '\u00FF' || length == 1 || !char.IsHighSurrogate(charA) || !char.IsHighSurrogate(charB))
+                char a = charA;
+                char b = charB;
+
+                if (!char.IsHighSurrogate(a) || index >= lengthA - 1 || !char.IsLowSurrogate(Unsafe.Add(ref charA, 1)))
```
Comment (safern) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/OrdinalCasing.Icu.cs:207]: Couldn't charA + 1 also change while checking if a is a high surrogate?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/OrdinalCasing.Icu.cs]: Yes. I have fixed it. thanks!

File: src/libraries/System.Private.CoreLib/src/System/Globalization/OrdinalCasing.Icu.cs [hunk 2]
```diff
@@ -196,72 +197,93 @@ internal static int CompareStringIgnoreCase(ref char strA, int lengthA, ref char
...
-                 // optimize for Ascii cases
-                if (charA <= '\u00FF' || length == 1 || !char.IsHighSurrogate(charA) || !char.IsHighSurrogate(charB))
+                char a = charA;
+                char b = charB;
+                char lowSurrogateA = '\0';
+
+                if (!char.IsHighSurrogate(a) || index >= lengthA - 1 || !char.IsLowSurrogate(lowSurrogateA = Unsafe.Add(ref charA, 1)))
                 {
-                    if (charA == charB)
+                    if (!char.IsHighSurrogate(b) || index >= lengthB - 1 || !char.IsLowSurrogate(Unsafe.Add(ref charB, 1)))
```
Comment (safern) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/OrdinalCasing.Icu.cs:210]: Missed this one, should we also store `lowSurrogateA`?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/OrdinalCasing.Icu.cs:210]: No, it is not needed here. We are not accessing `Unsafe.Add(ref charB, 1)` again inside such block.

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/OrdinalCasing.Icu.cs:210]: Ah, you are right 😄

**Review by safern — APPROVED** [review]
LGTM

**Comment by tarekgh** [comment]
Thanks @safern

**Pull request merged**